docs/src/overview.rst

```diff
@@ -4,27 +4,46 @@ Overview
 MLX C wraps and follows as closely as possible the C++ API of `MLX
 <https://github.com/ml-explore/mlx>`_.
 
-C Object
---------
-
-MLX C relies on several C ``struct`` which are all compatible with a private
-:doc:`object struct <object>` (that we will refer as a MLX C object),
-which implements basic memory management. When using MLX C, it is important
-to note that:
-
-* All returned C objects come with a reference count of 1, which needs to
-  be decreased with a matching :func:`mlx_free()` call.
-
-* When the reference count of an object gets to 0, the object is freed.
-
-* Reference count may be increased via :func:`mlx_retain()`, in which case it
-  should be matched at some point with a :func:`mlx_free()` to avoid any memory
-  leak. Increasing the reference count may be necessary when storing a
-  MLX C object into a custom C structure.
-
-In addition, MLX C objects come with a convenience :func:`mlx_tostring()`
-function which returns a string describing the contents of the object
-(which is particularly interesting for array objects).
+C Objects
+---------
+
+MLX C relies on several opaque C ``struct`` to operate. This includes:
+
+* :doc:`Arrays <array>`, on which computations are performed: :class:`mlx_array`.
+* :doc:`Devices <device>`, which define the compute unit where operations are performed: :class:`mlx_device`.
+* :doc:`Streams <stream>`, which ingest and execute array operations on a specific device: :class:`mlx_stream`.
+
+Other ``struct`` objects holding data, but not directly related to compute,
+are also available, such as :doc:`strings <string>`, :doc:`vectors
+<vector>` and :doc:`maps <map>`. There are also few extra objects such as
+:doc:`closures <closure>` (which encapsulate specific function signatures,
+possibly holding upvalues), and :doc:`distributed groups
+<distributed_group>` (which enable distributed computation).
+
+All these objects have in common the way they operate: one need to create
+them with a constructor functions, such as :func:`mlx_array_new()`, and one
+need to free the allocated object through a corresponding free call, for
+example :func:`mlx_array_free()`. There should be always one single
+``free`` call corresponding to a given ``new`` call.
+
+Once an object is created, one can perform multiple assignments, either
+through ``set`` functions (such as :func:`mlx_array_set()`), or through MLX
+operations. For example, the following code is perfectly valid in MLX C:
+
+.. code-block:: c
+
+    mlx_stream stream = mlx_gpu_stream();
```

Comment:
Does this need to be destroyed?  Maybe not since it isn't `new`?

Reply:
No, I think it needs to be freed -- the implementation of this is:

```
extern "C" mlx_stream mlx_gpu_stream() {
  try {
    return mlx_stream_new_(
```

Reply:
renamed as `mlx_default_gpu_stream_new()`